site/content/blog/2020-07-27-automated-ui-testing-with-cypress.md

```diff
@@ -0,0 +1,144 @@
+Before we dive into the guidelines and best practices, first let’s take a look at our E2E Test setup. The illustration below shows the overall test setup.
+
+![test environment image](/blog/2020-07-27-automated-ui-testing-with-cypress/test_environment_setup.png)
+
+Mattermost server and web application, which we'll refer to as "server", is spun up with all the required services such as PostgreSQL, Elasticsearch, SAML/Okta, OpenLDAP, Minio, Webhook server, Plugin Marketplace and Email server. Once the server is ready, Cypress will interact with it as a user and initiate actions and verify results. In some cases, it directly accesses the services to set up or verify information. There is also a separate static website generated from Storybook that is used to check the functionality of a component from the web application.
+
+## Typical Test Execution Life Cycle
+
+![test execution life cycle image](/blog/2020-07-27-automated-ui-testing-with-cypress/test_execution_life_cycle.png)
+
+The illustration above shows the test execution life cycle of our E2E tests. It starts with the initial setup where the test environment preparation happens such as spinning the test server. Then, testing each test file until all are completed. Finally, it consolidates each individual report and artifact, saves into AWS S3 and ReportPortal dashboard, and publishes a test summary to our community channel.
+
+## General Tips
+Now that we have an idea on how we set up and execute E2E tests with Cypress, let me share some of the guidelines and best practices we learned and formulated towards happy path for developers and contributors, including test scripts as they develop features and fix bugs, and for quality assurance (QA) team to easily capture regressions during nightly build and release testing. 
+
+### 1. Reset before test
+This is true both for server and user settings. In the past, we used to prepare test requirements in the test file only. During that time, changes in settings are easy to track and have minimal UI effect. However, it didn’t work well as we developed features and added more and more, and we were bitten many times by it. With that, we made sure in the global `before` hook ([source](https://github.com/mattermost/mattermost-webapp/blob/277a5cafac5385b1e283952a0881f459cffcbe94/e2e/cypress/support/index.js#L94-L164)) that the server and user, specifically sysadmin, are in a predetermined state before testing.
+
+### 2. Isolate test
+When using known test data like users, teams and channels, test cases are making changes to those test data and it is fine. However, it caused a lot of pain to prepare the state for the next test, so we decided to prevent sharing test data per test file by using a convenient custom command: `cy.apiInitSetup` ([source](https://github.com/mattermost/mattermost-webapp/blob/277a5cafac5385b1e283952a0881f459cffcbe94/e2e/cypress/support/api/setup.js#L4-L30)). Such command automatically creates test data with the typical use of:
```

Comment:
```suggestion
When using known test data like users, teams, and channels, test cases are making changes to those test data and it is fine. However, it caused a lot of pain to prepare the state for the next test, so we decided to prevent sharing test data per test file by using a convenient custom command: `cy.apiInitSetup` ([source](https://github.com/mattermost/mattermost-webapp/blob/277a5cafac5385b1e283952a0881f459cffcbe94/e2e/cypress/support/api/setup.js#L4-L30)). Such command automatically creates test data with the typical use of:
```